### PR TITLE
Update list command

### LINKED ISSUE #167: Allow list command to be able to list paired or unpaired persons

### PR DESCRIPTION
**Pull request by wz2k**
# Changes
* Can now list paired or unpaired persons
* more test cases

Resolves #167

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 1]
```diff
@@ -5,30 +5,155 @@
...
  */
 public class ListCommand extends Command {
 
     public static final String COMMAND_WORD = "list";
     public static final HashMap<Prefix, String> COMMAND_PROMPTS = new LinkedHashMap<>();
-    public static final String MESSAGE_SUCCESS = "Listed all elderly, volunteers and pairs";
+    public static final String MESSAGE_SUCCESS = "Listed all volunteers, elderly and pairs";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": List all volunteers and elderly pairs if nothing is"
+            + "specified after the command, else list all paired or unpaired volunteers and elderly if the"
+            + "\"paired\" or \"unpair\" (case-insensitive) word is given respectively after the the list command word"
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:30]:
```suggestion
            + "\"paired\" or \"unpaired\" (case-insensitive) word is given respectively after the the list command word"
```

`unpaired` or `unpair`?

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 2]
```diff
@@ -5,30 +5,155 @@
...
+            + "Parameters: [paired/unpair]\n"
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:31]: ditto

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 3]
```diff
@@ -5,30 +5,155 @@
...
+            + "Example: " + COMMAND_WORD + " paired";
+    public static final String MESSAGE_SUCCESS_LIST_PAIRED =
+            "Listed %1$s paired volunteer, %2$s paired elderly and all pairs";
+    public static final String MESSAGE_SUCCESS_LIST_UNPAIRED =
+            "Listed %1$s unpaired volunteer, %2$s unpaired elderly and all pairs";
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:36]: I think a better behaviour would be to not show any pairs at all when listing unpaired volunteers and elderly. What do you think?

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 4]
```diff
@@ -5,30 +5,155 @@
...
+        switch (action) {
+        case "":
+            model.refreshAllFilteredLists();
+            return new CommandResult(MESSAGE_SUCCESS);
+        case "unpaired":
+            model.updateFilteredElderlyList(getUnPairedElderlyPredicate(friendlyLink));
+            model.updateFilteredVolunteerList(getUnPairedVolunteerPredicate(friendlyLink));
+            model.updateFilteredPairList(showAllPairPredicate);
+            return new CommandResult(String.format(MESSAGE_SUCCESS_LIST_UNPAIRED,
+                    model.getFilteredVolunteerList().size(), model.getFilteredElderlyList().size()));
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:66]:
```suggestion
            model.refreshAllFilteredList(
                getUnPairedElderlyPredicate(friendlyLink),
                getUnPairedVolunteerPredicate(friendlyLink), 
                showAllPairPredicate);
            return new CommandResult(String.format(MESSAGE_SUCCESS_LIST_UNPAIRED,
                    model.getFilteredVolunteerList().size(), model.getFilteredElderlyList().size()));
```
Can you overload `refreshAllFilteredLists`function to take in these predicates instead?

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 5]
```diff
@@ -5,30 +5,155 @@
...
+     * @param friendlyLink Application cache.
+     * @return Elderly predicate.
+     */
+    public Predicate<Elderly> getUnPairedElderlyPredicate(FriendlyLink friendlyLink) {
+        requireNonNull(friendlyLink);
+        List<Pair> pairList = friendlyLink.getPairList();
+        List<Elderly> pairedElderlyList =
+                pairList.stream().map(pair -> pair.getElderly()).collect(Collectors.toList());
+
+        return elderly -> !pairedElderlyList.contains(elderly);
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:120]:
```suggestion
        requireNonNull(friendlyLink);
        return elderly -> !(getPairedElderlyPredicate(friendlyLink)(elderly));
```
Try if this works (same for volunteer)

File: src/main/java/seedu/address/logic/parser/ListCommandParser.java [hunk 6]
```diff
@@ -9,25 +12,36 @@
...
+     * @param map the ArgumentMultimap to be validated.
+     * @return true if the ArgumentMultimap is valid, false otherwise.
      */
-    @Override
-    public ListCommand parse(String userInput) throws ParseException {
-        assert false : "This method should not be invoked";
-        return null;
+    public static boolean validate(ArgumentMultimap map) {
+        String preamble = map.getPreamble().toLowerCase();
+        System.out.println(preamble);
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/parser/ListCommandParser.java:44]:
```suggestion

```

**Review by Zxun2 — APPROVED** [review]
lgtmbrsc 👍

File: src/main/java/seedu/address/logic/commands/ListCommand.java (hunk 4 above)
Comment (wz2k) [reply src/main/java/seedu/address/logic/commands/ListCommand.java:66]: Every predicate is different, the code will be worse.

File: src/main/java/seedu/address/logic/commands/ListCommand.java (hunk 5 above)
Comment (wz2k) [reply src/main/java/seedu/address/logic/commands/ListCommand.java:120]: It doesn't work. Thanks for trying anyways

File: src/main/java/seedu/address/logic/commands/ListCommand.java (hunk 3 above)
Comment (wz2k) [reply src/main/java/seedu/address/logic/commands/ListCommand.java:36]: List all pairs is better, there's no downside to it if they need to refer to the pair list will going through pair or unpaired people. On the other hand, not showing any pairs would by chance cause inconvenience if they need it.

Comment (Zxun2) [reply src/main/java/seedu/address/logic/commands/ListCommand.java:36]: Yup, make sense.

**Pull request merged**